### PR TITLE
[YUNIKORN-1555] Deduplicate async pod events during recovery

### PR DESCRIPTION
**Pull request by bgrams**
### What is this PR for?
Follow up to https://github.com/apache/yunikorn-k8shim/pull/532. Original PR failed to address duplicated events coming from the informer via `asyncEvent`s as this scenario was not covered by unit tests.

Changes involve passing state between recovery methods to deduplicate `AddPod` operations given the pod UID.

### What type of PR is it?
* [ ] - Bug Fix
* [ ] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [x] - Hot Fix
* [ ] - Refactoring

### What is the Jira issue?
[YUNIKORN-1555](https://issues.apache.org/jira/browse/YUNIKORN-1555)

### How should this be tested?
Unit tests modified

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/appmgmt/general/podevent_handler_test.go
Comment (bgrams) [inline pkg/appmgmt/general/podevent_handler_test.go:107]: I'd like to hear feedback on this assertion specifically, as the `UpdatePod` event simply completes the task without actually updating the associated pod. So the tests are passing but this is not extremely intuitive.

Comment (wilfred-s) [reply pkg/appmgmt/general/podevent_handler_test.go:107]: I think the assumption in the code that a UpdatePod event should transition the pod to finished is not correct. In the handling part of `Manager.updatePod` a check of the current pod and new pod is performed. Then it only transitions the pod if needed by calling the handler.

With VPA (vertical pod auto scaler) there might be more that we need to handle for updates, that is a discussion for later.

I think `PodEventHandler.updatePod` needs to get both pods and the logic of the `Manager.updatePod` needs to move to it. That needs a bit of "surgery" of the PodEventHandler, **BUT** not for as part of this change.

**Review by wilfred-s — CHANGES_REQUESTED** [review]

File: pkg/appmgmt/general/podevent_handler_test.go
Comment (pbacsko) [reply pkg/appmgmt/general/podevent_handler_test.go:107]: Created an umbrella (https://issues.apache.org/jira/browse/YUNIKORN-1765) and a subtask (https://issues.apache.org/jira/browse/YUNIKORN-1766) for this.

File: pkg/appmgmt/general/podevent_handler.go
Comment (bgrams) [reply pkg/appmgmt/general/podevent_handler.go:108]: I think the answer is no. However I chose not to implement the same check for updates/deletes given that this change is intended simply to delegate logic to the outer `recoverApps` method, which only covers adds. We can extend this to updates if need be.

Comment (bgrams) [reply pkg/appmgmt/general/podevent_handler.go:108]: to @pbacsko's point it might just be simpler to ditch resource versions for that reason

File: pkg/appmgmt/appmgmt_recovery.go
Comment (bgrams) [reply pkg/appmgmt/appmgmt_recovery.go:70]: I agree this makes more sense for the scope of this change. Updated.

File: pkg/appmgmt/appmgmt_recovery.go
Comment (pbacsko) [inline pkg/appmgmt/appmgmt_recovery.go:74]: This is updated outside of the "if" branch. Which might contain all kinds of pods not scheduled by Yunikorn.
We need to be inside the branch, plus, only store pods that are not running.

If a pod is running, you don't know whether we have a new event for it or not. You can't filter out events which are about an active pod.

**Review by pbacsko — CHANGES_REQUESTED** [review]
As I mentioned, you need to store finished pods. The `NeedRecovery()` condition has to be split up further.

**Review by pbacsko — APPROVED** [review]
+1 LGTM (pending tests)

cc @wilfred-s to get another review

**Review by craigcondit — APPROVED** [review]
Looks ok to me. I'd still like @wilfred-s to give the final sign-off as I think he has the best overall view of what's required here.

**Review by wilfred-s — APPROVED** [review]
LGTM

**Pull request closed without merging**
